server/src/main/java/io/spine/server/commandbus/CommandBus.java

```diff
@@ -178,37 +182,74 @@ protected CommandDispatcherRegistry createRegistry() {
 
     @SuppressWarnings("ReturnOfCollectionOrArrayField") // OK for a protected factory method
     @Override
-    protected Deque<BusFilter<CommandEnvelope>> createFilterChain() {
-        filterChain.push(scheduler);
-        return filterChain;
+    protected Collection<BusFilter<CommandEnvelope>> filterChainTail() {
+        return ImmutableList.of(scheduler);
+    }
+
+    @Override
+    protected Deque<BusFilter<CommandEnvelope>> filterChainHead() {
+        BusFilter<CommandEnvelope> tap = new CommandReceivedTap(systemGateway);
+        Deque<BusFilter<CommandEnvelope>> result = newLinkedList();
+        result.push(tap);
+        return result;
     }
 
     @Override
     protected CommandEnvelope toEnvelope(Command message) {
         return CommandEnvelope.of(message);
     }
 
+    @Override
+    protected StreamObserver<Ack> wrappedObserver(Iterable<Command> commands,
+                                                  StreamObserver<Ack> source) {
+        StreamObserver<Ack> wrappedSource = super.wrappedObserver(commands, source);
+        TenantId tenant = tenantOf(commands);
+        StreamObserver<Ack> result = CommandAcknowledgementMonitor
+                .newBuilder()
+                .setDelegate(wrappedSource)
+                .setTenantId(tenant)
+                .setSystemGateway(systemGateway)
+                .build();
+        return result;
+    }
+
+    private static TenantId tenantOf(Iterable<Command> commands) {
+        return Streams.stream(commands)
+                      .map(Commands::getTenantId)
+                      .findAny()
+                      .orElse(TenantId.getDefaultInstance());
+    }
+
     @Override
     protected void dispatch(CommandEnvelope envelope) {
         CommandDispatcher<?> dispatcher = getDispatcher(envelope);
+        onDispatchCommand(envelope);
         try {
             dispatcher.dispatch(envelope);
-            commandStore.setCommandStatusOk(envelope);
-        } catch (RuntimeException e) {
-            Throwable cause = getRootCause(e);
-            commandStore.updateCommandStatus(envelope, cause, log);
-            if (causedByRejection(e)) {
-                ThrowableMessage throwableMessage = (ThrowableMessage) cause;
-                Rejection rejection = toRejection(throwableMessage, envelope.getCommand());
-                Class<?> rejectionClass = AnyPacker.unpack(rejection.getMessage())
-                                                   .getClass();
-                Log.log()
-                   .trace("Posting rejection {} to RejectionBus.", rejectionClass.getName());
-                rejectionBus().post(rejection);
-            }
+        } catch (RuntimeException exception) {
+            errorHandler.handleError(envelope, exception);
         }
     }
 
+    private void onDispatchCommand(CommandEnvelope command) {
```

Comment:
I think it makes sense to group these `on...` methods somehow. They aren't looking good being standalone.